woocommerce/woocommerce
# Prevent synchronous WCCOM API calls in admin pages

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/51082.

Currently, two endpoints are making synchronous API calls that could cause performance issues and potential fatal errors:

1. `OnboardingThemes.php` - Makes synchronous API calls on wc-admin pages
2. `WCPayPromotion.php` - Makes synchronous API calls on all wp-admin pages

These synchronous calls can lead to poor admin performance and fatal errors if the requests fail or timeout.

This PR removes the `OnboardingThemes.php` class and modifies `WCPayPromotion.php` to use cached specs instead of live API calls (similar to recent PR #50674). It also limits the `isWooPayEligible` setting to only be added on Payments-related pages.

The `OnboardingThemes.php` class was introduced for the old setup wizard themes step. It is in the internal namespace and is not used in the core or our extensions. I checked the WP directory and didn't find any plugins or themes that use this class either, so it should be safe to remove. Please help double-check this. 🙏 

The `isWooPayEligible` setting was introduced in https://github.com/woocommerce/woocommerce/pull/39596 WCPay banners. It should be limited to ~Payments-related pages~ WC admin and Payments page as it is relevant only there. 



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### API Calls shouldn't crash the site

1. Create a fresh site
2. Enable marketing suggestions in `WooCommerce > Settings > Advanced > WooCommerce.com`
3. Add throw `new \Exception();` in [this line](https://github.com/woocommerce/woocommerce/blob/20d89bae03397094eac80a5f24fedb3b48c95645/plugins/woocommerce/src/Internal/Admin/WCPayPromotion/Init.php#L167)
4. Clear transients: `wp transient delete --all && wp cache flush`
5. Visit WP Admin and WC Admin pages (except payment settings page)
6. Verify page loads without fatal errors

#### Test WooPay Promotion

1. Remove the exception from the previous test
2. Set store country to US
3. Go to `/wp-admin/admin.php?page=wc-admin&path=%2Fwc-pay-welcome-page` and open the browser console
4. Verify that `window.wcSettings.admin.isWooPayEligible` is `true`
5. Confirm it mentions `(including WooPay merchant terms)` in the banner

![Screenshot 2024-10-24 at 12 02 31](https://github.com/user-attachments/assets/5fea5370-dc05-4a85-8100-34a6a9a4d07c)

6. Go to `/wp-admin/admin.php?page=wc-settings&tab=checkout`
7. Verify that `window.wcSettings.admin.isWooPayEligible` is `true` and it should show a description mentioning WooPay.

![Screenshot 2024-10-24 at 11 27 57](https://github.com/user-attachments/assets/5762549b-c9e4-4c2c-8055-20a1ad9b84e3)

8. Install and Activate WooPayments plugin
9. Reload the `WooCommerce > Settings > Payments` page, it should WooPay icon in the banner


![Screenshot 2024-10-24 at 11 30 25](https://github.com/user-attachments/assets/8d0fe46b-2e35-4722-80c6-838c64468857)

10. Change store country to `Afghanistan`
11. Go to `/wp-admin/admin.php?page=wc-admin&task=payments` and verify that `window.wcSettings.admin.isWooPayEligible` is `false`





### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
